[FEATURE] Jwt 로그아웃 로직 작성

JSY8869 opened the pull request:
1. 로그아웃 handler 구현

로그아웃 과정에서 쿠키 내에 있는 토큰을 삭제하고 SecurityContext 정보 초기화 하여야 하기 때문에 해당 로직 구현.
access 토큰은 redis 를 사용하지 않기 때문에 클라이언트가 헤더에서 직접 삭제하도록 구현.

2. 로그아웃 성공 handler 구현

로그아웃 성공 시 재로그인을 위해 로그인 화면으로 이동할 수 있도록 redirect를 요청하는 응답을 보냄

3. security configuration에 로그아웃 설정 추가

deleteCookies 메서드 -> default 옵션이 달라서 현재 프로젝트의 토큰 속성과 달라 토큰이 삭제되지 않아서 삭제 후 직접 구현
(심지어 속성 설정이 불가능함)
clearAuthentication 메서드 -> default가 true이기 때문에 삭제

4. TODO 주석 삭제
#24 
- USERNOTFOUND 예외가 다른 곳에서도 사용되는 메서드에서 던져지기 때문에 공통적으로 사용되는 예외를 상속 받도록 결정

5. code refactoring

삭제용 쿠키 객체 싱글톤 형식으로 변경

JSY8869 committed:
feat: Add jwt logout handler

1. 로그아웃 handler 구현
로그아웃 과정에서 쿠키 내에 있는 토큰을 삭제하고 SecurityContext 정보 초기화 하여야 하기 때문에 해당 로직 구현.
access 토큰은 redis 를 사용하지 않기 때문에 클라이언트가 헤더에서 직접 삭제하도록 구현.

2. 로그아웃 성공 handler 구현
로그아웃 성공 시 재로그인을 위해 로그인 화면으로 이동할 수 있도록 redirect를 요청하는 응답을 보냄

3. security configuration에 로그아웃 설정 추가
deleteCookies 메서드 -> 옵션이 달라서 토큰이 삭제되지 않아서 삭제 후 직접 구현
clearAuthentication 메서드 -> default가 true이기 때문에 삭제

related to #27

JSY8869 committed:
comment: Deleting comment because the problem is solved

1. 문제가 해결되어 Todo 주석 삭제
UserNotFoundException이 여러 곳에서 공통적으로 사용되는 로직에서 던져지는 예외이기 때문에 IllegalArgumentException 를 상속 받는 것으로 결정

resolves #24

JSY8869 committed:
refact: code refactoring

1. 삭제용 cookie 생성 로직 변경
생성된 쿠키가 모두 같은 속성을 가지기 때문에 매번 쿠키를 생성하는 것이 아닌 하나의 삭제용 쿠키를 만들어서 사용하는 방식으로 변경

2. refactoring
로그아웃 성공 메시지가 불필요하다고 판단하여 삭제
end point를 상수로 선언하여 사용하도록 변경

related to #27

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/security/handler/logout/JwtLogoutHandler.java:
```diff
@@ -8,14 +8,17 @@
-        cookie.setMaxAge(0);
-        cookie.setHttpOnly(true);
-        response.addCookie(cookie);
+        response.addCookie(COOKIE);
```
쿠키를 삭제하지 않고 왜 만료기간이 0인 쿠키를 만드는건가요?

JSY8869 replied:
쿠키를 따로 삭제하는 것이 불가능 하여 같은 이름과 속성을 가진 쿠키를 만들어서 바로 만료되도록 만료기간을 0으로 설정하여 사용자에게 전달하는 방식으로 쿠키를 삭제시켰습니다.

CokeLee777 reviewed (commented):
리뷰보고 커맨트 남겨주세요~

JSY8869 committed:
refact: Refactoring cookie

1. logout handler 쿠키 생성 메서드 생성
생성자에서 로직을 바로 실행하는 것이 아닌 메서드로 분할하여 해당 메서드를 통해 삭제용 쿠키의 설정을 진행하도록 변경

2. 쿠키 경로 설정
모든 곳에서 쿠키 사용이 가능하도록 쿠키의 경로를 "/" 으로 설정

related to #27

CokeLee777 reviewed (approved):
머지해도 될 것 같습니다~

JSY8869 committed:
Merge branch 'feature/authorization' into feature/27
